### ISSUE TITLE
Fine tuning the motion-cosegmentation Model 

### ISSUE THREAD

**Issue by SaiThejeshwar** [AliaksandrSiarohin/motion-cosegmentation#38]
Hi,

This is very interesting work.

I wanted to fine tune the model with few more new videos.  Though I could do the video-processing part, I am unable to run the training code. I am getting the below error. How should i fine tune the model.
I have followed the steps in the repository. Please help me out!

# !python train.py --config "config/Retrain_15segments.yaml" --checkpoint "models/vox-cpk.pth.tar"

![image](https://user-images.githubusercontent.com/18008582/110251294-daac7a00-7fa5-11eb-8875-a2b0ef23489e.png)

**Comment by AliaksandrSiarohin** [comment]
Checkpoint and config should have the same number of segments (keypoints) in your case it is 10 and 15 respectively. Either use different checkpoint or different config.

**Comment by SaiThejeshwar** [comment]
Thank you for the reply.

Yes. I have already tried different number of segments, and it worked for 10 segments and checkpoint is saved in log folder. But, When trying to infer, with the new fine tuned checkpoint model weights, I get the missing key error in the checkpoint - "blend_downsample.weight" (in Image). Please guide me.

![image](https://user-images.githubusercontent.com/18008582/110270528-09e5da00-7fec-11eb-8d08-35d6c570859f.png)

The above is thrown, when I exceute this block during Inference.

reconstruction_module, segmentation_module = load_checkpoints(config='config/Retrain_10segments.yaml', 
                                               checkpoint='log/Retrain_10segments 08-03-21 03:38:02/00000009-checkpoint.pth.tar',
                                               blend_scale=0.125, first_order_motion_model=False)


The configuration file, I have adapted is from the "vox-256-sem-10segments.yaml" with slight modifications .

Below is my config file.

------
# Dataset parameters
dataset_params:
  # Path to data, data can be stored in several formats: .mp4 or .gif videos, stacked .png images or folders with frames.
  root_dir: video-preprocessing/croppednew23dataset
  # Image shape, needed for staked .png format.
  image_shape: [256, 256, 3]
  # In case of VoxCeleb or TaiChi single video can be splitted in many chunks, or the maybe several videos for single person.
  # In this case epoch can be a pass over different videos (if id_sampling=True) or over different chunks (if id_sampling=False)
  # If the name of the video '12335#adsbf.mp4' the id is assumed to be 12335
  id_sampling: False
  # Augmentation parameters see augmentation.py for all posible augmentations
  augmentation_params:
    flip_param:
      horizontal_flip: True
      time_flip: True
    jitter_param:
      brightness: 0.1
      contrast: 0.1
      saturation: 0.1
      hue: 0.1

# Defines model architecture
model_params:
  common_params:
    # Number of segments
    num_segments: 10
    # Number of channels per image
    num_channels: 3
    # Use only shift and no-affine part
    estimate_affine_part: True
  segmentation_module_params:
     # Softmax temperature for shift heatmaps
     temperature: 0.1
     # Number of features mutliplier
     block_expansion: 32
     # Maximum allowed number of features
     max_features: 1024
     # Number of block in Unet. Can be increased or decreased depending or resolution.
     num_blocks: 5
     # Segmentations is predicted on smaller images for better performance,
     # scale_factor=0.25 means that 256x256 image will be resized to 64x64
     scale_factor: 0.25
  reconstruction_module_params:
    # Number of features mutliplier
    block_expansion: 64
    # Maximum allowed number of features
    max_features: 512
    # Number of downsampling blocks in Jonson architecture.
    # Can be increased or decreased depending or resolution.
    num_down_blocks: 2
    # Number of ResBlocks  in Jonson architecture.
    num_bottleneck_blocks: 6
    # Use visibility map or not
    estimate_visibility: True

# Parameters of training
train_params:
  num_workers: 5
  # Number of training epochs
  num_epochs: 5
  # For better i/o performance when number of videos is small number of epochs can be multiplied by this number.
  # Thus effectively with num_repeats=100 each epoch is 100 times larger.
  num_repeats: 2
  # Learning rates
  lr_segmentation_module: 2.0e-4
  lr_reconstruction_module: 2.0e-4
  batch_size: 4
  # Scales for perceptual pyramide loss. If scales = [1, 0.5, 0.25, 0.125] and image resolution is 256x256,
  # than the loss will be computer on resolutions 256x256, 128x128, 64x64, 32x32.
  scales: [1, 0.5, 0.25, 0.125]
  # Save checkpoint this frequently. If checkpoint_freq=50, checkpoint will be saved every 50 epochs.
  checkpoint_freq: 50
  # Parameters of transform for equivariance loss
  transform_params:
    # Sigma for affine part
    sigma_affine: 0.05
    # Sigma for deformation part
    sigma_tps: 0.005
    # Number of point in the deformation grid
    points_tps: 5
  loss_weights:
    equivariance: 10
    perceptual: [10, 10, 10, 10, 10]


# Visualization parameters
visualizer_params:
  # Draw keypoints (shifts in affine transformations) of this size, increase or decrease depending on resolution
  kp_size: 5
  # Draw white border around images
  draw_border: True
  # Color map for keypoints
  colormap: 'gist_rainbow'

------

Thank you

**Comment by AliaksandrSiarohin** [comment]
You are using ```--supervised ``` flag? It is not clear to me, what you are trying to achieve. If the goal is to use model in supervised mode you should finetune fomm.

**Comment by SaiThejeshwar** [comment]
> You are using `--supervised ` flag? It is not clear to me, what you are trying to achieve. If the goal is to use model in supervised mode you should finetune fomm.

My goal is to swap faces in the video. So, which one should I fine tune.

**Comment by AliaksandrSiarohin** [comment]
https://github.com/AliaksandrSiarohin/first-order-model

**Comment by SaiThejeshwar** [comment]
> https://github.com/AliaksandrSiarohin/first-order-model

Do you mean, I need to fine tune fomm?

To be more specific, I am aiming at this - https://github.com/AliaksandrSiarohin/first-order-model#face-swap

**Comment by AliaksandrSiarohin** [comment]
Yes you need to finetune fomm than.

**Comment by abelghazinyan** [comment]
Hi @AliaksandrSiarohin thanks for amazing work!

I was trying to fine-tune motion-cosegmentation model with my dataset but same problem occurred.

So at first I tried to fine-tune fomm model with 10 segments config file, the training went without problems, but when I load the models it gives me
![image](https://user-images.githubusercontent.com/26743458/130321454-0fd8ae5e-69f8-4ecc-acff-01b2fea93546.png)

Then I tried to fine-tune motion-copart's checkpoint with 15 segments. The training went without problems, but when trying to test it I got the same error as above.

My goal is to fine-tune motion-co parts checkpoint with my dataset and then run it with supervised option, without fomm and using face parsing.
What is missing?

**Comment by abelghazinyan** [comment]
So quick update on my problem

This error occurs when I run with --supervised flag.
But when I run your pertained model vox-15segments.pth.tar it works.

But when I fine-tune that model or train it from scratch with vox-256-sem-15segments.yaml config it only works without --supervised flag.

So my question is why your checkpoint works with supervised flag but mine not?

**Comment by AliaksandrSiarohin** [comment]
Supervised uses original fomm, you should not fine-tune.

**Comment by abelghazinyan** [comment]
But I can use supervised with your vox-15segments.pth.tar and I guess it's motion-co part model not fomm.
So you mean I should fine-tune fomm to use it with supervised flag in motion-co part?
